### ISSUE TITLE
index error in Dn.rn() if it is called with an empty Dn

### ISSUE THREAD

**Issue by mtimm** [datacenter/cobra#46]
On python 3 unit tests for request.py are failing when `ConfigRequest.getRootMo()` is called when there is a fvTenant and a fvBD added to the ConfigRequest object.  The basic code path that leads to this issue looks like this:

```
import cobra.model.fv
import cobra.mit.request
fvTenant = cobra.model.fv.Tenant('uni', 'testing')
fvBD = cobra.model.fv.BD('uni/tn-testing', 'test')
cr = cobra.mit.request.ConfigRequest()
cr.addMo(fvBD)
cr.addMo(fvTenant)
cr.getRootMo()
```

The error is:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mtimm/.venv/test_python3/lib/python3.4/site-packages/acicobra-0.1-py3.4.egg/cobra/mit/request.py", line 1018, in getRootMo
    rootMo)
  File "/Users/mtimm/.venv/test_python3/lib/python3.4/site-packages/acicobra-0.1-py3.4.egg/cobra/mit/request.py", line 1064, in __getMoForDnInFlatTree
    return flatTree.setdefault(dn, mo if mo else ConfigRequest.__makeMoFromDn(dn))
  File "/Users/mtimm/.venv/test_python3/lib/python3.4/site-packages/acicobra-0.1-py3.4.egg/cobra/mit/request.py", line 1069, in __makeMoFromDn
    namingVals = list(dn.rn().namingVals)
  File "/Users/mtimm/.venv/test_python3/lib/python3.4/site-packages/acicobra-0.1-py3.4.egg/cobra/mit/naming.py", line 344, in rn
    return self.__rns[-1]
IndexError: list index out of range
```

The `ConfigRequest.getRootMo()` call ends up in `Dn.findCommonParent()` which can return an empty Dn:

```
        if index == 0:
            return Dn()
```

Then later in `ConfigRequest.getRootMo()` there is a call to `ConfigRequest.__getMoForDnInFlatTree` which can call `ConfigRequest.__makeMoFromDn()` which has the call to `Dn.rn()`:

```
namingVals = list(dn.rn().namingVals)
```

I need to investigate why this does not fail in this way in Python 2.  In python 2.7.8 the following is returned from getRootMo in this case:

```
>>> cr.getRootMo()
<cobra.model.fv.Tenant object at 0x10ee88710>
>>> 
```

**Comment by mtimm** [comment]
This occurs because **cmp** is not used by python 3 (it is literally ignored) and there is no cmp builtin in python 3.  This leads to the following code in `cobra.mit.naming.findCommonParent` to return False incorrectly only in python 3:

```
def allRnsEqual(allDns, i):
    firstRn = None
    for eachDn in allDns:
        currentRn = eachDn.__rns[i]
        if firstRn is None:
            firstRn = currentRn
        else:
            if firstRn != currentRn:
                return False

    # All Rns at this level are equal
    return True
```

We do use this so at least cmp will exist:

```
from past.builtins import cmp
```

So the last piece to get this part of the code fuctioning for Python 3 is to add the rich comparison operators for Rn's - `__lt__`, `__le__`, `__eq__`, `__ne__`, `__gt__` and `__ge__`.  

I'll have to go through all of the classes with custom **cmp**'s and see if they all require these rich comparison operators.

**Comment by mtimm** [comment]
Closed by #47
